src/vcpkg/base/messages.cpp

```diff
@@ -0,0 +1,283 @@
+                std::abort();
+            }
+            ptr += written;
+            to_write -= written;
+        }
+    }
+
+    void write_unlocalized_text_to_stdout(Color c, StringView sv)
+    {
+        static constexpr char reset_color_sequence[] = {'\033', '[', '0', 'm'};
+
+        if (sv.empty()) return;
+
+        bool reset_color = false;
+        if (::isatty(STDOUT_FILENO) && c != Color::None)
+        {
+            reset_color = true;
+
+            const char set_color_sequence[] = {'\033', '[', '9', static_cast<char>(c), 'm'};
+            write_all(STDOUT_FILENO, set_color_sequence, sizeof(set_color_sequence));
+        }
+
+        write_all(STDOUT_FILENO, sv.data(), sv.size());
+
+        if (reset_color)
+        {
+            write_all(STDOUT_FILENO, reset_color_sequence, sizeof(reset_color_sequence));
+        }
+    }
+#endif
+
+    namespace
+    {
+        struct Messages
+        {
+            // this is basically a SoA - each index is:
+            // {
+            //   name
+            //   default_string
+            //   localization_comment
+            //   localized_string
+            // }
+            // requires: names.size() == default_strings.size() == localized_strings.size()
+            std::vector<StringLiteral> names;
+            std::vector<StringLiteral> default_strings;       // const after startup
+            std::vector<StringLiteral> localization_comments; // const after startup
+
+            bool initialized = false;
+            std::vector<std::string> localized_strings;
+        };
+
+        // to avoid static initialization order issues,
+        // everything that needs the messages needs to get it from this function
+        Messages& messages()
+        {
+            static Messages m;
+            return m;
+        }
+    }
+
+    void threadunsafe_initialize_context()
+    {
+        Messages& m = messages();
+        if (m.initialized)
+        {
+            write_unlocalized_text_to_stdout(Color::Error,
+                                 "double-initialized message context; this is a very serious bug in vcpkg\n");
+            Checks::exit_fail(VCPKG_LINE_INFO);
+        }
+        m.localized_strings.resize(m.names.size());
+        m.initialized = true;
+    }
+    void threadunsafe_initialize_context(const Json::Object& message_map)
+    {
+        threadunsafe_initialize_context();
+
+        Messages& m = messages();
+        std::vector<std::string> names_without_localization;
+
+        for (::size_t index = 0; index < m.names.size(); ++index)
+        {
+            const auto& name = m.names[index];
+            if (auto p = message_map.get(m.names[index]))
+            {
+                m.localized_strings[index] = p->string().to_string();
+            }
+            else if (Debug::g_debugging)
+            {
+                // we only want to print these in debug
+                names_without_localization.push_back(name);
```

Comment:
I don't think `--debug` should immediately imply "never localize anything". Maybe there is a need for a separate `--debug-locale` or some other method of getting this output.

Reply:
this is a misunderstanding of how this works.